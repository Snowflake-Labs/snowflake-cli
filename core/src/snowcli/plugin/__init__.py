from dataclasses import dataclass
from enum import Enum
from typing import List, Optional

from snowcli.plugin.api import PluginCommandGroupSpec


class PluginLoadingMode(Enum):
    ALL_INSTALLED_PLUGINS = 1
    ONLY_ENABLED_PLUGINS = 2


@dataclass
class PluginPackageInfo:
    package_name: str
    package_version: str
    dependencies: List[str]


@dataclass
class LoadedPlugin:
    plugin_name: str
    is_enabled: bool
    plugin_package_info: PluginPackageInfo
    command_group_spec: PluginCommandGroupSpec


@dataclass
class PluginAfterVerification:
    plugin: LoadedPlugin
    verification_error: Optional[Exception]

    def is_ok(self) -> bool:
        return self.verification_error is None

    def error_str(self) -> Optional[str]:
        error = self.verification_error
        if error:
            return f"{error.__class__.__name__} - {error.__str__()}"
        else:
            return None
