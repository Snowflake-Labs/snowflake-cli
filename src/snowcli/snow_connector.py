from __future__ import annotations

import os
from io import StringIO
from pathlib import Path
from typing import Optional

import click
import snowflake.connector
from jinja2 import Environment, FileSystemLoader
from snowflake.connector.cursor import SnowflakeCursor

TEMPLATES_PATH = Path(__file__).parent / "sql"


def get_standard_stage_name(name: str) -> str:
    # Handle embedded stages
    if name.startswith("snow://"):
        return name

    return f"@{name}"


class SnowflakeConnector:
    """Initialize a connection from a snowsql-formatted config"""

    def __init__(
        self,
        connection_config: dict,
        overrides: Optional[dict] = None,
    ):
        self.connection_config = connection_config
        if overrides:
            for config, value in ((k, v) for k, v in overrides.items() if v):
                self.connection_config[config] = value
        self.connection_config["application"] = self._find_command_path()
        self.ctx = snowflake.connector.connect(**self.connection_config)
        self.cs = self.ctx.cursor()

    @staticmethod
    def _find_command_path():
        ctx = click.get_current_context(silent=True)
        if ctx:
            # Example: SNOWCLI.WAREHOUSE.STATUS
            return ".".join(["SNOWCLI", *ctx.command_path.split(" ")[1:]]).upper()
        return "SNOWCLI"

    def __del__(self):
        try:
            self.cs.close()
            self.ctx.close()
        except (TypeError, AttributeError):
            pass

    def get_version(self):
        self.cs.execute("SELECT current_version()")
        return self.cs.fetchone()[0]

    def create_function(
        self,
        name: str,
        input_parameters: str,
        return_type: str,
        handler: str,
        imports: str,
        database: str,
        schema: str,
        role: str,
        warehouse: str,
        overwrite: bool,
        packages: list[str],
    ) -> SnowflakeCursor:
        return self.run_sql(
            "create_function",
            {
                "database": database,
                "schema": schema,
                "role": role,
                "warehouse": warehouse,
                "name": name,
                "overwrite": overwrite,
                "input_parameters": input_parameters,
                "return_type": return_type,
                "handler": handler,
                "imports": imports,
                "packages": packages,
                "signature": self.generate_signature_from_params(
                    input_parameters,
                ),
            },
        )

    def create_procedure(
        self,
        name: str,
        input_parameters: str,
        return_type: str,
        handler: str,
        imports: str,
        database: str,
        schema: str,
        role: str,
        warehouse: str,
        overwrite: bool,
        packages: list[str],
        execute_as_caller: bool,
    ) -> SnowflakeCursor:
        return self.run_sql(
            "create_procedure",
            {
                "database": database,
                "schema": schema,
                "role": role,
                "warehouse": warehouse,
                "name": name,
                "overwrite": overwrite,
                "input_parameters": input_parameters,
                "return_type": return_type,
                "handler": handler,
                "imports": imports,
                "packages": packages,
                "signature": self.generate_signature_from_params(
                    input_parameters,
                ),
                "execute_as_caller": execute_as_caller,
            },
        )

    def upload_file_to_stage(
        self,
        file_path,
        destination_stage,
        path,
        role,
        database,
        warehouse,
        schema,
        overwrite,
        parallel: int = 4,
        create_stage: bool = True,
    ):
        create_stage_command = ""
        if create_stage:
            create_stage_command = (
                f"create stage if not exists {destination_stage} "
                "comment='deployments managed by snowcli'"
            )

        full_stage_name = (
            f"@{destination_stage}"
            if not destination_stage.startswith("snow://")
            else destination_stage
        )
        return self.run_sql(
            "put_stage",
            {
                "role": role,
                "database": database,
                "schema": schema,
                "warehouse": warehouse,
                "path": file_path,
                "destination_path": path,
                "name": full_stage_name,
                "create_stage_command": create_stage_command,
                "parallel": parallel,
                "overwrite": overwrite,
            },
        )

    def execute_function(
        self,
        function,
        database,
        schema,
        role,
        warehouse,
    ) -> SnowflakeCursor:
        return self.run_sql(
            "execute_function",
            {
                "database": database,
                "schema": schema,
                "role": role,
                "warehouse": warehouse,
                "function": function,
            },
        )

    def execute_procedure(
        self,
        procedure,
        database,
        schema,
        role,
        warehouse,
    ) -> SnowflakeCursor:
        return self.run_sql(
            "call_procedure",
            {
                "database": database,
                "schema": schema,
                "role": role,
                "warehouse": warehouse,
                "procedure": procedure,
            },
        )

    def describe_function(
        self,
        database,
        schema,
        role,
        warehouse,
        signature=None,
        name=None,
        input_parameters=None,
        show_exceptions=True,
    ) -> SnowflakeCursor:
        if signature is None and name and input_parameters:
            signature = name + self.generate_signature_from_params(input_parameters)
        return self.run_sql(
            "describe_function",
            {
                "database": database,
                "schema": schema,
                "role": role,
                "warehouse": warehouse,
                "signature": signature,
            },
            show_exceptions,
        )

    def describe_procedure(
        self,
        database,
        schema,
        role,
        warehouse,
        signature=None,
        name=None,
        input_parameters=None,
        show_exceptions=True,
    ) -> SnowflakeCursor:
        if signature is None and name and input_parameters:
            signature = name + self.generate_signature_from_params(input_parameters)
        return self.run_sql(
            "describe_procedure",
            {
                "database": database,
                "schema": schema,
                "role": role,
                "warehouse": warehouse,
                "signature": signature,
            },
            show_exceptions,
        )

    def list_functions(
        self,
        database,
        schema,
        role,
        warehouse,
        like="%%",
    ) -> SnowflakeCursor:
        return self.run_sql(
            "list_functions",
            {
                "database": database,
                "schema": schema,
                "role": role,
                "warehouse": warehouse,
                "like": like,
            },
        )

    def list_stages(
        self,
        database,
        schema,
        role,
        warehouse,
        like="%%",
    ) -> SnowflakeCursor:
        return self.run_sql(
            "list_stages",
            {
                "database": database,
                "schema": schema,
                "role": role,
                "warehouse": warehouse,
                "like": like,
            },
        )

    def list_stage(
        self,
        database,
        schema,
        role,
        warehouse,
        name,
        like="%%",
    ) -> SnowflakeCursor:
        name = get_standard_stage_name(name)

        return self.run_sql(
            "list_stage",
            {
                "database": database,
                "schema": schema,
                "role": role,
                "warehouse": warehouse,
                "name": name,
                "like": like,
            },
        )

    def get_stage(
        self,
        database,
        schema,
        role,
        warehouse,
        name,
        path,
    ) -> SnowflakeCursor:
        name = get_standard_stage_name(name)

        return self.run_sql(
            "get_stage",
            {
                "database": database,
                "schema": schema,
                "role": role,
                "warehouse": warehouse,
                "name": name,
                "path": path,
            },
        )

    def set_procedure_comment(
        self,
        database,
        schema,
        role,
        warehouse,
        signature=None,
        name=None,
        input_parameters=None,
        show_exceptions=True,
        comment="",
    ) -> SnowflakeCursor:
        if signature is None and name and input_parameters:
            signature = name + self.generate_signature_from_params(input_parameters)
        return self.run_sql(
            "set_procedure_comment",
            {
                "database": database,
                "schema": schema,
                "role": role,
                "warehouse": warehouse,
                "signature": signature,
                "comment": comment,
            },
            show_exceptions,
        )

    def put_stage(
        self,
        database,
        schema,
        warehouse,
        role,
        name,
        path,
        overwrite: bool = False,
        parallel: int = 4,
    ) -> SnowflakeCursor:
        name = get_standard_stage_name(name)

        return self.run_sql(
            "put_stage",
            {
                "database": database,
                "schema": schema,
                "role": role,
                "warehouse": warehouse,
                "name": name,
                "path": path,
                "overwrite": overwrite,
                "parallel": parallel,
            },
        )

    def remove_from_stage(
        self, database, schema, role, warehouse, name, path
    ) -> SnowflakeCursor:
        name = get_standard_stage_name(name)

        return self.run_sql(
            "remove_from_stage",
            {
                "database": database,
                "schema": schema,
                "role": role,
                "warehouse": warehouse,
                "name": name,
                "path": path,
            },
        )

    def create_stage(
        self,
        database,
        schema,
        role,
        warehouse,
        name,
    ) -> SnowflakeCursor:
        return self.run_sql(
            "create_stage",
            {
                "database": database,
                "schema": schema,
                "role": role,
                "warehouse": warehouse,
                "name": name,
            },
        )

    def drop_stage(
        self,
        database,
        schema,
        role,
        warehouse,
        name,
    ) -> SnowflakeCursor:
        return self.run_sql(
            "drop_stage",
            {
                "database": database,
                "schema": schema,
                "role": role,
                "warehouse": warehouse,
                "name": name,
            },
        )

    def list_procedures(
        self,
        database,
        schema,
        role,
        warehouse,
        like="%%",
    ) -> SnowflakeCursor:
        return self.run_sql(
            "list_procedures",
            {
                "database": database,
                "schema": schema,
                "role": role,
                "warehouse": warehouse,
                "like": like,
            },
        )

    def drop_function(
        self,
        database,
        schema,
        role,
        warehouse,
        signature,
    ) -> SnowflakeCursor:
        return self.run_sql(
            "drop_function",
            {
                "database": database,
                "schema": schema,
                "role": role,
                "warehouse": warehouse,
                "signature": signature,
            },
        )

    def drop_procedure(
        self,
        database,
        schema,
        role,
        warehouse,
        signature,
    ) -> SnowflakeCursor:
        return self.run_sql(
            "drop_procedure",
            {
                "database": database,
                "schema": schema,
                "role": role,
                "warehouse": warehouse,
                "signature": signature,
            },
        )

    def list_streamlits(
        self,
        database="",
        schema="",
        role="",
        warehouse="",
    ) -> SnowflakeCursor:
        return self.run_sql(
            "list_streamlits",
            {
                "database": database,
                "schema": schema,
                "role": role,
                "warehouse": warehouse,
            },
        )

    def show_warehouses(
        self,
        database="",
        schema="",
        role="",
        warehouse="",
        like="%%",
    ) -> SnowflakeCursor:
        return self.run_sql(
            "show_warehouses",
            {
                "database": database,
                "schema": schema,
                "role": role,
                "warehouse": warehouse,
            },
        )

    def create_streamlit(
        self,
        database="",
        schema="",
        role="",
        warehouse="",
        name="",
        file="",
        from_stage_command="",
    ) -> SnowflakeCursor:
        return self.run_sql(
            "create_streamlit",
            {
                "database": database,
                "schema": schema,
                "role": role,
                "warehouse": warehouse,
                "name": name,
                "file_name": file,
                "from_stage_command": from_stage_command,
            },
        )

    def share_streamlit(
        self,
        database="",
        schema="",
        role="",
        warehouse="",
        name="",
        to_role="",
    ) -> SnowflakeCursor:
        return self.run_sql(
            "share_streamlit",
            {
                "database": database,
                "schema": schema,
                "role": role,
                "warehouse": warehouse,
                "name": name,
                "to_role": to_role,
            },
        )

    def drop_streamlit(
        self,
        database="",
        schema="",
        role="",
        warehouse="",
        name="",
    ) -> SnowflakeCursor:
        return self.run_sql(
            "drop_streamlit",
            {
                "database": database,
                "schema": schema,
                "role": role,
                "warehouse": warehouse,
                "name": name,
            },
        )

    def deploy_streamlit(
        self,
        name,
        file_path,
        stage_path,
        role,
        database,
        schema,
        warehouse,
        overwrite,
    ):
        stage_name = f"snow://streamlit/{database}.{schema}.{name}/default_checkout"

        """Upload main python file to stage and return url of streamlit"""
        self.upload_file_to_stage(
            file_path=file_path,
            destination_stage=stage_name,
            path=stage_path,
            role=role,
            database=database,
            schema=schema,
            warehouse=warehouse,
            overwrite=overwrite,
            create_stage=False,
        )

        result = self.run_sql(
            "get_streamlit_url",
            {
                "name": name,
                "role": role,
                "database": database,
                "schema": schema,
            },
        )

        return result.fetchone()[0]

    def describe_streamlit(self, name, database, schema, role, warehouse):
        description = self.run_sql(
            "describe_streamlit",
            {
                "name": name,
                "role": role,
                "database": database,
                "schema": schema,
                "warehouse": warehouse,
            },
        )
        url = self.run_sql(
            "get_streamlit_url",
            {
                "name": name,
                "role": role,
                "database": database,
                "schema": schema,
            },
        )
        return (description, url)

    def run_sql(
        self,
        command,
        context,
        show_exceptions=True,
    ) -> SnowflakeCursor:
        env = Environment(loader=FileSystemLoader(TEMPLATES_PATH))
        template = env.get_template(f"{command}.sql")
        sql = template.render(**context)
        try:
            if os.getenv("DEBUG"):
                print(f"Executing sql:\n{sql}")
            results = self.ctx.execute_stream(StringIO(sql))

            # Return result from last cursor
            *_, last_result = results
            return last_result
        except snowflake.connector.errors.ProgrammingError as e:
            if show_exceptions:
                print(f"Error executing sql:\n{sql}")
                print(e)
            raise (e)

    @staticmethod
    def generate_signature_from_params(params: str) -> str:
        if params == "()":
            return "()"
        return "(" + " ".join(params.split()[1::2])
