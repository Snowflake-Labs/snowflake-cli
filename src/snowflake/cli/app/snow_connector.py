from __future__ import annotations

import contextlib
import logging
import os
from typing import Dict, Optional

import click
import snowflake.connector
from click.core import ParameterSource  # type: ignore
from click.exceptions import ClickException
from snowflake.cli.api.config import get_connection, get_default_connection
from snowflake.cli.api.exceptions import (
    InvalidConnectionConfiguration,
    SnowflakeConnectionError,
)
from snowflake.cli.app.telemetry import command_info
from snowflake.connector import SnowflakeConnection
from snowflake.connector.errors import DatabaseError, ForbiddenError

log = logging.getLogger(__name__)

ENCRYPTED_PKCS8_PK_HEADER = b"-----BEGIN ENCRYPTED PRIVATE KEY-----"
UNENCRYPTED_PKCS8_PK_HEADER = b"-----BEGIN PRIVATE KEY-----"


def connect_to_snowflake(
    temporary_connection: bool = False,
    mfa_passcode: Optional[str] = None,
    connection_name: Optional[str] = None,
    **overrides,
) -> SnowflakeConnection:
    connection_parameters = _get_connection_details(
        connection_name, temporary_connection, overrides
    )

    connection_parameters = _update_connection_details_with_private_key(
        connection_parameters
    )

    if mfa_passcode:
        connection_parameters["passcode"] = mfa_passcode

    try:
        # Whatever output is generated when creating connection,
        # we don't want it in our output. This is particularly important
        # for cases when external browser and json format are used.
        with contextlib.redirect_stdout(None):
            return snowflake.connector.connect(
                application=command_info(),
                **connection_parameters,
            )
    except ForbiddenError as err:
        raise SnowflakeConnectionError(err)
    except DatabaseError as err:
        raise InvalidConnectionConfiguration(err.msg)


def _get_connection_details(
    connection_name: str | None = None,
    temporary_connection: bool = False,
    overrides: Dict | None = None,
):
    if not temporary_connection:
        if connection_name is not None:
            connection_parameters = (
                get_connection(connection_name)
                if connection_name
                else get_default_connection()
            )
        else:
            connection_parameters = get_default_connection()
    else:
        connection_parameters = {}
    if overrides:
        ctx = click.get_current_context(silent=True)
        for k, v in overrides.items():
            if v is None:
                continue
            # Apply override if:
            # 1. There is not context
            # 2. There is context and override source is a flag
            # 3. There is a context and override is from flag envvar but
            #    the key is not present in connection details from connection
            if not ctx or (
                ctx.get_parameter_source(k) != ParameterSource.ENVIRONMENT  # type: ignore
                or k not in connection_parameters
            ):
                connection_parameters[k] = v
    return connection_parameters


def _update_connection_details_with_private_key(connection_parameters: Dict):
    if "private_key_path" in connection_parameters:
        if connection_parameters.get("authenticator") == "SNOWFLAKE_JWT":
            private_key = _load_pem_to_der(connection_parameters["private_key_path"])
            connection_parameters["private_key"] = private_key
            del connection_parameters["private_key_path"]
        else:
            raise ClickException(
                "Private Key authentication requires authenticator set to SNOWFLAKE_JWT"
            )
    return connection_parameters


def _load_pem_to_der(private_key_path: str) -> bytes:
    """
    Given a private key file path (in PEM format), decode key data into DER
    format
    """
    from cryptography.hazmat.backends import default_backend
    from cryptography.hazmat.primitives.serialization import (
        Encoding,
        NoEncryption,
        PrivateFormat,
        load_pem_private_key,
    )

    with open(private_key_path, "rb") as f:
        private_key_pem = f.read()

    private_key_passphrase = os.getenv("PRIVATE_KEY_PASSPHRASE", None)
    if (
        private_key_pem.startswith(ENCRYPTED_PKCS8_PK_HEADER)
        and private_key_passphrase is None
    ):
        raise ClickException(
            "Encrypted private key, you must provide the"
            "passphrase in the environment variable PRIVATE_KEY_PASSPHRASE"
        )

    if not private_key_pem.startswith(
        ENCRYPTED_PKCS8_PK_HEADER
    ) and not private_key_pem.startswith(UNENCRYPTED_PKCS8_PK_HEADER):
        raise ClickException(
            "Private key provided is not in PKCS#8 format. Please use correct format."
        )

    if private_key_pem.startswith(UNENCRYPTED_PKCS8_PK_HEADER):
        private_key_passphrase = None

    private_key = load_pem_private_key(
        private_key_pem,
        str.encode(private_key_passphrase)
        if private_key_passphrase is not None
        else private_key_passphrase,
        default_backend(),
    )

    return private_key.private_bytes(
        encoding=Encoding.DER,
        format=PrivateFormat.PKCS8,
        encryption_algorithm=NoEncryption(),
    )
