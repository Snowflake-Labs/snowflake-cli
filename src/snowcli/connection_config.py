from __future__ import annotations

import os
from configparser import ConfigParser
from pathlib import Path
from snowcli.exception import EnvironmentVariableNotFoundError


class ConnectionConfigs:
    def __init__(self, snowsql_config_path="~/.snowsql/config"):
        self.snowsql_config_path = snowsql_config_path
        self._config = ConfigParser(inline_comment_prefixes="#", interpolation=None)
        p = Path(snowsql_config_path).absolute().expanduser()
        self._config.read(p)

    def is_connection_exists(self, connection_name: str):
        return self._config.sections().__contains__(f"connections.{connection_name}")

    def get_connections(self):
        connections_names = [
            n for n in self._config.sections() if n.startswith("connections.")
        ]
        return {n: self._get_connection(n) for n in connections_names}

    def get_connection(self, connection_name: str):
        return self._get_connection(f"connections.{connection_name}")

    def add_connection(self, connection_name: str, entry: dict):
        self._config[f"connections.{connection_name}"] = {}
        connection = self._config[f"connections.{connection_name}"]
        for (k, v) in entry.items():
            connection[k] = v

        with open(os.path.expanduser(self.snowsql_config_path), "w+") as f:
            self._config.write(f)

    def _get_connection(self, connection_full_name: str):
        connection = self._config[connection_full_name]
        # Remap names to appropriate args in Python Connector API
        connection = {
            k.replace("name", ""): v.strip('"') for k, v in connection.items()
        }
        return self._replace_with_env_variables(connection)

    def _replace_with_env_variables(self, connection_sections: dict):
        for (name, value) in connection_sections.items():
            if value.startswith("$SNOWCLI_"):
                try:
                    connection_sections[name] = os.environ[value[1:]]
                except (KeyError):
                    raise EnvironmentVariableNotFoundError(value[1:]) from None
        return connection_sections
