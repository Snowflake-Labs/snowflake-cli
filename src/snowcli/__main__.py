from __future__ import annotations

import sys

from snowcli.cli.app import app


if __name__ == "__main__":
    app()

if getattr(sys, "frozen", False):
    app(sys.argv[1:])
