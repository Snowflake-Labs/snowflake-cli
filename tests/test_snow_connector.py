import os
import textwrap
from unittest import mock

import pytest

from snowcli.snow_connector import SnowflakeConnector, SnowflakeCursor


# Used as a solution to syrupy having some problems with comparing multilines string
class CustomStr(str):
    def __repr__(self):
        return str(self)


MOCK_CONNECTION = {
    "database": "databaseValue",
    "schema": "schemaValue",
    "role": "roleValue",
    "warehouse": "warehouseValue",
}


@pytest.mark.parametrize(
    "cmd,expected",
    [
        (["sql", "-q", "foo"], "SNOWCLI.SQL"),
        (["warehouse", "status"], "SNOWCLI.WAREHOUSE.STATUS"),
    ],
)
@mock.patch("snowcli.snow_connector.snowflake.connector.connect")
def test_command_context_is_passed_to_snowflake_connection(
    mock_conn, runner, cmd, expected
):
    mock_cursor = mock.Mock(spec=SnowflakeCursor)
    mock_cursor.description = []
    mock_cursor.fetchall.return_value = []

    mock_conn.return_value.execute_stream.return_value = [mock_cursor]
    mock_conn.return_value.execute_string.return_value = [mock_cursor]
    result = runner.invoke_with_config(cmd)
    assert result.exit_code == 0, result.output
    kwargs = mock_conn.call_args_list[-1][-1]
    assert "application" in kwargs
    assert kwargs["application"] == expected


@mock.patch("snowflake.connector")
def test_create_procedure(_, snapshot):
    connector = SnowflakeConnector(connection_parameters=MOCK_CONNECTION)
    connector.ctx.execute_stream.return_value = (None, None)

    connector.create_procedure(
        name="nameValue",
        input_parameters="(string a, variant b)",
        return_type="returnTypeValue",
        handler="handlerValue",
        imports="import1, import2",
        database="databaseValue",
        schema="schemaValue",
        role="roleValue",
        warehouse="warehouseValue",
        overwrite=True,
        packages=["aaa", "bbb"],
        execute_as_caller=True,
    )
    query, *_ = connector.ctx.execute_stream.call_args.args
    assert query.getvalue() == snapshot


@mock.patch("snowflake.connector")
def test_execute_procedure(_, snapshot):
    connector = SnowflakeConnector(connection_parameters=MOCK_CONNECTION)
    connector.ctx.execute_stream.return_value = (None, None)

    connector.execute_procedure(
        procedure="procedureValue",
        database="databaseValue",
        schema="schemaValue",
        role="roleValue",
        warehouse="warehouseValue",
    )
    query, *_ = connector.ctx.execute_stream.call_args.args
    assert query.getvalue() == snapshot


@mock.patch("snowflake.connector")
def test_describe_procedure(_, snapshot):
    connector = SnowflakeConnector(connection_parameters=MOCK_CONNECTION)
    connector.ctx.execute_stream.return_value = (None, None)

    connector.describe_procedure(
        database="databaseValue",
        schema="schemaValue",
        role="roleValue",
        warehouse="warehouseValue",
        signature="signatureValue",
        name="nameValue",
        input_parameters="(string a, variant b)",
        show_exceptions="show_exceptionsValue",
    )
    query, *_ = connector.ctx.execute_stream.call_args.args
    assert query.getvalue() == snapshot


@mock.patch("snowflake.connector")
def test_set_procedure_comment(_, snapshot):
    connector = SnowflakeConnector(connection_parameters=MOCK_CONNECTION)
    connector.ctx.execute_stream.return_value = (None, None)

    connector.set_procedure_comment(
        database="databaseValue",
        schema="schemaValue",
        role="roleValue",
        warehouse="warehouseValue",
        signature="signatureValue",
        name="nameValue",
        input_parameters="(string a, variant b)",
        show_exceptions="show_exceptionsValue",
        comment="commentValue",
    )
    query, *_ = connector.ctx.execute_stream.call_args.args
    assert query.getvalue() == snapshot


@mock.patch("snowflake.connector")
def test_list_procedures(_, snapshot):
    connector = SnowflakeConnector(connection_parameters=MOCK_CONNECTION)
    connector.ctx.execute_stream.return_value = (None, None)

    connector.list_procedures(
        database="databaseValue",
        schema="schemaValue",
        role="roleValue",
        warehouse="warehouseValue",
        like="likeValue",
    )
    query, *_ = connector.ctx.execute_stream.call_args.args
    assert query.getvalue() == snapshot


@mock.patch("snowflake.connector")
def test_drop_procedure(_, snapshot):
    connector = SnowflakeConnector(connection_parameters=MOCK_CONNECTION)
    connector.ctx.execute_stream.return_value = (None, None)

    connector.drop_procedure(
        database="databaseValue",
        schema="schemaValue",
        role="roleValue",
        warehouse="warehouseValue",
        signature="signatureValue",
    )
    query, *_ = connector.ctx.execute_stream.call_args.args
    assert query.getvalue() == snapshot


@mock.patch("snowflake.connector")
def test_list_streamlits(_, snapshot):
    connector = SnowflakeConnector(connection_parameters=MOCK_CONNECTION)
    connector.ctx.execute_stream.return_value = (None, None)

    connector.list_streamlits(
        database="databaseValue",
        schema="schemaValue",
        role="roleValue",
        warehouse="warehouseValue",
    )
    query, *_ = connector.ctx.execute_stream.call_args.args
    assert query.getvalue() == snapshot


@mock.patch("snowflake.connector")
def test_create_streamlit(_, snapshot):
    connector = SnowflakeConnector(connection_parameters=MOCK_CONNECTION)
    connector.ctx.execute_stream.return_value = (None, None)

    connector.create_streamlit(
        database="databaseValue",
        schema="schemaValue",
        role="roleValue",
        warehouse="warehouseValue",
        name="nameValue",
        file="fileValue",
    )
    query, *_ = connector.ctx.execute_stream.call_args.args
    assert query.getvalue() == snapshot


@mock.patch("snowflake.connector")
def test_create_streamlit_from_stage(_, snapshot):
    connector = SnowflakeConnector(connection_parameters=MOCK_CONNECTION)
    connector.ctx.execute_stream.return_value = (None, None)

    connector.create_streamlit(
        database="databaseValue",
        schema="schemaValue",
        role="roleValue",
        warehouse="warehouseValue",
        name="nameValue",
        file="fileValue",
        from_stage_command="FROM @stageValue",
    )
    query, *_ = connector.ctx.execute_stream.call_args.args
    assert query.getvalue() == snapshot


@mock.patch("snowflake.connector")
def test_share_streamlit(_, snapshot):
    connector = SnowflakeConnector(connection_parameters=MOCK_CONNECTION)
    connector.ctx.execute_stream.return_value = (None, None)

    connector.share_streamlit(
        database="databaseValue",
        schema="schemaValue",
        role="roleValue",
        warehouse="warehouseValue",
        name="nameValue",
        to_role="to_roleValue",
    )
    query, *_ = connector.ctx.execute_stream.call_args.args
    assert query.getvalue() == snapshot


@mock.patch("snowflake.connector")
def test_drop_streamlit(_, snapshot):
    connector = SnowflakeConnector(connection_parameters=MOCK_CONNECTION)
    connector.ctx.execute_stream.return_value = (None, None)

    connector.drop_streamlit(
        database="databaseValue",
        schema="schemaValue",
        role="roleValue",
        warehouse="warehouseValue",
        name="nameValue",
    )
    query, *_ = connector.ctx.execute_stream.call_args.args
    assert query.getvalue() == snapshot


@mock.patch("snowflake.connector")
def test_deploy_streamlit(_, snapshot):
    connector = SnowflakeConnector(connection_parameters=MOCK_CONNECTION)
    connector.ctx.execute_stream.return_value = (mock.MagicMock(),)
    connector.upload_file_to_stage = mock.MagicMock()

    connector.deploy_streamlit(
        name="nameValue",
        file_path="file_pathValue",
        stage_path="stage_pathValue",
        role="roleValue",
        database="databaseValue",
        schema="schemaValue",
        warehouse="warehouseValue",
        overwrite=True,
    )
    query, *_ = connector.ctx.execute_stream.call_args.args
    assert query.getvalue() == snapshot


@pytest.mark.parametrize(
    "create_stage",
    [True, False],
)
@pytest.mark.parametrize("stage_name", ["namedStageValue", "snow://embeddedStageValue"])
@mock.patch("snowflake.connector")
def test_upload_file_to_stage(_, snapshot, create_stage, stage_name):
    connector = SnowflakeConnector(connection_parameters=MOCK_CONNECTION)
    connector.ctx.execute_stream.return_value = (None, None)

    connector.upload_file_to_stage(
        file_path="file_pathValue",
        destination_stage=stage_name,
        path="pathValue",
        role="roleValue",
        database="databaseValue",
        schema="schemaValue",
        warehouse="warehouseValue",
        overwrite="overwriteValue",
        create_stage=create_stage,
    )
    query, *_ = connector.ctx.execute_stream.call_args.args
    assert query.getvalue() == snapshot


@mock.patch("snowflake.connector")
def test_describe_streamlit(_, snapshot):
    connector = SnowflakeConnector(connection_parameters=MOCK_CONNECTION)
    connector.ctx.execute_stream.return_value = (None, None)

    connector.describe_streamlit(
        name="nameValue",
        database="databaseValue",
        schema="schemaValue",
        role="roleValue",
        warehouse="warehouseValue",
    )
    query, *_ = connector.ctx.execute_stream.call_args.args
    assert query.getvalue() == snapshot


@mock.patch("snowcli.cli.snowpark.registry.connect_to_snowflake")
def test_registry_get_token(mock_conn, runner):
    mock_conn.return_value.ctx._rest._token_request.return_value = {
        "data": {
            "sessionToken": "token1234",
            "validityInSecondsST": 42,
        }
    }
    result = runner.invoke(["snowpark", "registry", "token"])
    assert result.exit_code == 0, result.output
    assert result.stdout == '{"token": "token1234", "expires_in": 42}'


@mock.patch.dict(os.environ, {}, clear=True)
def test_returns_nice_error_in_case_of_connectivity_error(runner):
    result = runner.invoke_with_config(["sql", "-q", "select 1"])
    assert result.exit_code == 1
    assert "Invalid connection configuration" in result.output
    assert "User is empty" in result.output
