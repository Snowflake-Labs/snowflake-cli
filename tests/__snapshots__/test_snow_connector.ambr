# serializer version: 1
# name: test_create_procedure
  '''
  use role roleValue;
  use warehouse warehouseValue;
  use database databaseValue;
  use schema schemaValue;
  CREATE OR REPLACE  PROCEDURE nameValue(string a, variant b)
           RETURNS returnTypeValue
           LANGUAGE PYTHON
           RUNTIME_VERSION=3.8
           IMPORTS=('import1, import2')
           HANDLER='handlerValue'
           PACKAGES=('aaa','bbb')
           EXECUTE AS CALLER;
  
  
  describe PROCEDURE nameValue(a, b);
  '''
# ---
# name: test_describe_procedure
  '''
  use role roleValue;
  use warehouse warehouseValue;
  use database databaseValue;
  use schema schemaValue;
  desc PROCEDURE signatureValue;
  '''
# ---
# name: test_set_procedure_comment
  '''
  use role roleValue;
  use warehouse warehouseValue;
  use database databaseValue;
  use schema schemaValue;
  alter PROCEDURE signatureValue SET COMMENT = $$commentValue$$;
  '''
# ---
# name: test_upload_file_to_stage[namedStageValue-False]
  '''
  use role roleValue;
  use warehouse warehouseValue;
  use database databaseValue;
  use schema schemaValue;
  
  
  
  put file://file_pathValue @namedStageValuepathValue auto_compress=false parallel=4 overwrite=overwriteValue;
  '''
# ---
# name: test_upload_file_to_stage[namedStageValue-True]
  '''
  use role roleValue;
  use warehouse warehouseValue;
  use database databaseValue;
  use schema schemaValue;
  
  
  create stage if not exists namedStageValue comment='deployments managed by snowcli';
  
  
  put file://file_pathValue @namedStageValuepathValue auto_compress=false parallel=4 overwrite=overwriteValue;
  '''
# ---
# name: test_upload_file_to_stage[snow://embeddedStageValue-False]
  '''
  use role roleValue;
  use warehouse warehouseValue;
  use database databaseValue;
  use schema schemaValue;
  
  
  
  put file://file_pathValue snow://embeddedStageValuepathValue auto_compress=false parallel=4 overwrite=overwriteValue;
  '''
# ---
# name: test_upload_file_to_stage[snow://embeddedStageValue-True]
  '''
  use role roleValue;
  use warehouse warehouseValue;
  use database databaseValue;
  use schema schemaValue;
  
  
  create stage if not exists snow://embeddedStageValue comment='deployments managed by snowcli';
  
  
  put file://file_pathValue snow://embeddedStageValuepathValue auto_compress=false parallel=4 overwrite=overwriteValue;
  '''
# ---
